src/models/trigger/block.ts

```diff
@@ -0,0 +1,40 @@
+import * as avs_pb from "../../../grpc_codegen/avs_pb";
+import Trigger, { TriggerProps, TriggerTypes } from "./interface";
+
+// Required props for constructor: id, name, type and data: { interval }
+export type BlockTriggerDataType = avs_pb.BlockCondition.AsObject;
+export type BlockTriggerProps = TriggerProps & { data: BlockTriggerDataType };
+
+class BlockTrigger extends Trigger {
+  constructor(props: BlockTriggerProps) {
+    super({ ...props, type: TriggerTypes.BLOCK, data: props.data });
+  }
+
+  toRequest(): avs_pb.TaskTrigger {
+    const request = new avs_pb.TaskTrigger();
+    request.setName(this.name);
+
+    if (!this.data) {
+      throw new Error(`Trigger data is missing for ${this.type}`);
+    }
+
+    const condition = new avs_pb.BlockCondition();
+    condition.setInterval((this.data as BlockTriggerDataType).interval);
+    request.setBlock(condition);
+
+    return request;
+  }
+
+  static fromResponse(raw: avs_pb.TaskTrigger): BlockTrigger {
+    // Convert the raw object to TriggerProps, which should keep name and id
+    const obj = raw.toObject() as unknown as TriggerProps;
+
+    return new BlockTrigger({
+      ...obj,
+      type: TriggerTypes.BLOCK,
```

Comment:
no need to pass `type`. you will notice that in `toRequest` we dont need to send type to server.

and when you're in `BlockTrigger` class, you definetely know its type. no need to set this redundant param

Reply:
I agree, we can use a `typeof BlockTrigger` to tell the type, but that’s not ideal, due to the below two reasons:
1. The result of the `typeof` can’t directly map to the original type’s values, such as `TriggerTypes.BLOCK`.
2. The `typeof` is not conventional and a little hacky, and we rarely use it.

I think the reason protobuff doesn’t need a `type` variable is that it does not support class inheritance, and use `oneOf` instead. However, the fact that there are 5 optional parameters, `epochs`,`cron`, `conditions`, etc. , using `oneOf` on a single struct, really confused me. As a caller, I didn’t know what are the right parameters. 

Additionally, no matter what trigger type a caller chooses, the data field is all called `data`. I also kept the inner variable original, so we know that the inner data is a `cron`, `epochList`, or `conditionsList`.

Reply:
i mean is there a reason we couldn't do this?

```
class Trigger {
  getType(): string {
     // parse class name to get type
    return this.constructor.name.substr(0, this.constructor.name.length - 7);
  }
}


class FooTrigger extends Trigger {
}

class BarTrigger extends Trigger {
}

const a = new FooTrigger();
console.log("im", a.getType());

const b = new BarTrigger();
console.log("im", b.getType());
```

Result

```
 npx tsc a.ts
❯ node a.js
im Foo
im Bar
```

it save users the work of manually passing the type parameter. BlockTrigger is definetely a block, just by infering the class name as i show above

Reply:
Same reason as above. There’s no FooTrigger can be or should be defined as a type, except for the original protobuf values.

```
export enum TriggerTypeCase {
        TRIGGER_TYPE_NOT_SET = 0,
        MANUAL = 2,
        FIXED_TIME = 3,
        CRON = 4,
        BLOCK = 5,
        EVENT = 6,
    }
```